### DISCUSSION TITLE
ESP32s/ESP-WROOM-32 with SD and LAN8720 ethernet

### DISCUSSION THREAD

**Discussion by Terrevue** [forkineye/ESPixelStick#689] [Q&A] [answered]
I have this ESP32s board:
![image](https://github.com/forkineye/ESPixelStick/assets/45849973/08da4571-0639-45f4-ac2a-6652ef91186a)

I have added an SD board and LAN8720 board in accordance with this layout I found in this thread: https://github.com/forkineye/ESPixelStick/issues/268

![image](https://github.com/forkineye/ESPixelStick/assets/45849973/709e0c90-9ab8-4d2a-aa36-c7b8ce26d4df)

As a result, here is how I have the boards configured:
![image](https://github.com/forkineye/ESPixelStick/assets/45849973/b75d18d1-c1d3-47d4-9162-8f19d4fdeb83)

When I attempt to boot after installing the latest ESP3DEUXQUAD_DMX image (the only one I could find that matched the pin configuration), I receive the following:

17:41:48: [EthDrv] Entering State: Powering Up
17:41:48: [ 12336][E][esp32-hal-gpio.c:95] __pinMode(): Invalid pin selected
17:41:48: E (12331) gpio: gpio_set_level(226): GPIO output gpio_num error
17:41:48: E (13198) esp_eth: esp_eth_ioctl(348): ethernet driver handle can't be null
17:41:49: [EthDrv] Entering State: Connecting
17:41:49: [EthDrv] ETH IP Before Start: 0.0.0.0
17:41:49: [EthDrv] Ethernet Connecting as ESP32001
17:41:50: [EthDrv] Connecting to Ethernet using DHCP
17:41:50: [EthDrv] Entering State: Wait for IP Address

My question is twofold:

1. Is there something obvious I am missing here?
2. OR, is there a different pinout/hardware type I should try?

Thanks in advance!

Edit:
One thing I haven't done is pulled high/low in accordance with this diagram:
![image](https://github.com/forkineye/ESPixelStick/assets/45849973/1df6b3f3-53aa-4d2e-8d28-05833be326de)

It was my understanding based on issue #268 that was no longer needed.  If I am incorrect, please let me know.

**Comment by MartinMueller2003** [comment]
The GPIOs used for Ethernet cannot be changed. You have to use the GPIOs as defined in the platform definitions files EXACTLY. The hardware is very specific about how this gets wired.

Here are the GPIO definitions.

#define ETH_ADDR_PHY_TLK110            31
#define DEFAULT_ETH_ADDR               ETH_ADDR_PHY_LAN8720
#define DEFAULT_ETH_TXEN               gpio_num_t::GPIO_NUM_21
#define DEFAULT_ETH_TXD0               gpio_num_t::GPIO_NUM_19
#define DEFAULT_ETH_TXD1               gpio_num_t::GPIO_NUM_22
#define DEFAULT_ETH_CRSDV              gpio_num_t::GPIO_NUM_27
#define DEFAULT_ETH_RXD0               gpio_num_t::GPIO_NUM_25
#define DEFAULT_ETH_RXD1               gpio_num_t::GPIO_NUM_26
#define DEFAULT_ETH_MDC_PIN            gpio_num_t::GPIO_NUM_23
#define DEFAULT_ETH_MDIO_PIN           gpio_num_t::GPIO_NUM_18

**Reply by Terrevue** [reply]
My apologies.  It appears that last graphic was misleading.  That was only included to show how there is a recommendation that nINIT be pulled high and NC be pulled low.  

My pin configuration was off by one pin by mistake.  Corrected and now Ethernet is connected.  However, I am still getting an "Invalid pin selected" message:

19:47:57: [EthDrv] Entering State: Powering Up
19:48:05: [ 12030][E][esp32-hal-gpio.c:95] __pinMode(): Invalid pin selected
19:48:05: E (12025) gpio: gpio_set_level(226): GPIO output gpio_num error

Not sure if that is a result of the Power Control Pin error in the UI:

![image](https://github.com/forkineye/ESPixelStick/assets/45849973/3f767026-26c0-4826-aec4-ccb0b1db17ad)


A few follow up questions:

1. Is the DEUX platform the best to be using in my case?  For example, I notice that FPP isn't an option for a secondary input.
2. Once ethernet is connected and verified, should WIFI be disabled? 
3. I'm getting repeated "Connected with IP" messages (see below).  Is this normal behavior?

19:49:50: [EthDrv] Entering State: Connecting
19:51:21: [EthDrv] Entering State: Got IP
19:51:21: [EthDrv] Ethernet Connected with IP: 192.168.150.118
19:51:21: [EthDrv] Entering State: Connecting
19:53:44: [EthDrv] Entering State: Got IP
19:53:45: [EthDrv] Ethernet Connected with IP: 192.168.150.118
19:53:45: [EthDrv] Entering State: Connecting
19:54:40: [EthDrv] Entering State: Got IP
19:54:41: [EthDrv] Ethernet Connected with IP: 192.168.150.118

Thanks for the assistance!

**Comment by MartinMueller2003** [comment]
You should not be getting repeated connect messages. There may be some noise in your HW setup. The signals going to the Eth board are pretty sensitive. 

The LoLin ESP32 - Pro board may be a better starting point for you. You would need to change the SD card GPIO settings but the Eth settings remain the same. 

Can you send me a link to the CPU board? I can add it to the list of platforms in January. FYI: None of the compiled versions are set up for the 32S version. There are known differences that could be causing the issue you are seeing.

**Reply by Terrevue** [reply]
Thanks for the response.  Here is a link to the board:
https://www.amazon.com/Teyleten-Robot-ESP-WROOM-32-Development-Microcontroller/dp/B08246MCL5?th=1

I've cloned the repository and was attempting to make the appropriate modifications myself.  Although I know C++, this is my first time working with this platform so I'm not expecting to actually make it work.  From what I have gathered, it appears the changes need to be made in a new GPIO_defs_[board name].hpp file as well as in GPIO_defs.hpp and platformio.ini.  I'll give it a shot and if it works, I'll submit a PR.

As for the ethernet card:

1. Any recommendations on what may be causing noise?  I know that's a broad ask, but I'm hoping there are a few common areas I should check first
2. Is there any advantage to pulling nINIT high and NC low, or has the code changes made that irrelevant?

Thanks again for your help!

**Comment by MartinMueller2003** [comment]
I usually pull NC pins high through a 4.7K resistor. nInt should be pulled high. It controls how the lock input is used.

**Reply by Terrevue** [reply]
Just to clarify, you're pulling both NC and nINT high?  Based on this discussion, the recommendation is to pull NC low.  Just wanted to double check before trying it, in case there was a typo.

https://github.com/espressif/arduino-esp32/issues/2907

**Comment by MartinMueller2003** [comment]
NC means - No Connection. In general it is not good to have a pin floating (even those not connected to an internal substrate) so designers will pull them high or low (whatever is most convenient). I prefer high. Others prefer low. both will work on an NC pin.

nINT is a configuration strapping pin and to work in this configuration it must be pulled high.

**Reply by Terrevue** [reply]
Thanks for the clarification!

**Comment by MartinMueller2003** [comment, marked as answer]
You will also need to add an entry in the Platformio.ini file. In that file you will add the build information for your device. Use the LoLin Pro as your guide. Change the board = ??? to select the CPU you are using. That setting controls a lot of the libraries used for the freertos HW support.

**Reply by Terrevue** [reply]
Working through the build now.  Thanks for all of the assistance!
